nemo_aligner/algorithms/generation.py

```diff
@@ -0,0 +1,508 @@
+            
+            if (done and buffer) or sum(
+                [len(b["prompts_only"]) for b in buffer]
+            ) == self.rollout_micro_batch_size:
+                
+                #candidate_responses = [[] for _ in range(sum([len(b["prompt_lengths"]) for b in buffer]))]
+                #candidate_responses = []
+                gen_tokens, prompt_lens, gen_lens, valids = [], [], [], []
+                for _ in range(self.num_responses_to_gen):
+                    # Generation happens on GPU but returned tensors are on CPU so as not to blow up VRAM due to self.num_responses_to_gen
+                    gen_tokens_buf, gen_prompt_lengths_buf, gen_lengths_buf, is_end = self.get_generations(buffer)
+                    #candidate_responses.append((gen_tokens_buf, gen_prompt_lengths_buf, gen_lengths_buf, is_end))
+                    gen_tokens.append(gen_tokens_buf)
+                    prompt_lens.append(gen_prompt_lengths_buf)
+                    gen_lens.append(gen_lengths_buf)
+                    valids.append(is_end)
+                    
+                    #for idx, (t, s, e, end) in enumerate(zip(gen_tokens_buf, gen_prompt_lengths_buf.tolist(), gen_lengths_buf.tolist(), is_end.tolist())):
+                    #    candidate_responses[idx].append((t,s,e,end))
+                
+                #final_batch = []
+                #for cand_list in candidate_responses:
+                #    prompt_only = cand_list[0][0][:cand_list[0][1]]
+                #    resp_list = batch_pad_to_fixed_len([s[0][s[1]:s[2]] for s in cand_list], max_batch_len, pad_token=self.model.tokenizer.eos_id)
+                
+                new_batch = {
+                    "prompt_and_response": torch.stack([batch_pad_to_fixed_len(x, self.model.cfg.encoder_seq_length, pad_token=self.model.tokenizer.eos_id) for x in gen_tokens], dim=0).cuda(non_blocking=True),
+                    "prompt_lens": torch.stack(prompt_lens, dim=0).cuda(non_blocking=True),
+                    "gen_lens": torch.stack(gen_lens, dim=0).cuda(non_blocking=True),
+                    "valids": torch.stack(valids, dim=0).cuda(non_blocking=True),
+                }
+                
+                yield new_batch
+                del new_batch, gen_tokens, prompt_lens, gen_lens, valids
+
+                buffer.clear()
+    '''
+
+    def augment_dataloader(self, dataloader):
+        """Augment dataloader with generations"""
+        iter_dataloader = iter(dataloader)
+        while True:
+            try:
+                batches = next(iter_dataloader)
+                batch = generate_sft_custom_collate(
+                    batches,
+                    eos_id=self.model.tokenizer.eos_id,
+                    reset_position_ids=self.model.cfg.data.get("reset_position_ids", False),
+                    reset_attention_mask=self.model.cfg.data.get("reset_attention_mask", False),
+                    eod_mask_loss=self.model.cfg.data.get("eod_mask_loss", False),
```

Comment:
Those flags are actually ignored in `generate_sft_custom_collate()` => is there any reason to keep them around?